File: Q47/Q47.ts
// Perform matrix multiplication.

function multiplyMatrices(
  matrixA: number[][],
  matrixB: number[][]
): number[][] | null {
  const rowsA = matrixA.length;
  const colsA = matrixA[0].length;
  const rowsB = matrixB.length;
  const colsB = matrixB[0].length;

  // Check if matrices can be multiplied
  if (colsA !== rowsB) {
    console.error("Matrices cannot be multiplied: incompatible dimensions.");
    return null;
  }

  // Initialize result matrix with zeros
  const result: number[][] = Array(rowsA)
    .fill(0)
    .map(() => Array(colsB).fill(0));

  // Perform multiplication
  for (let i = 0; i < rowsA; i++) {
    for (let j = 0; j < colsB; j++) {
      for (let k = 0; k < colsA; k++) {
        result[i][j] += matrixA[i][k] * matrixB[k][j];
      }
    }
  }

  return result;
}

const matrixA = [
  [1, 2, 3],
  [4, 5, 6],
];
const matrixB = [
  [7, 8],
  [9, 10],
  [11, 12],
];

const result = multiplyMatrices(matrixA, matrixB);
console.log(result);

File: Q47/Q47.py
# Perform matrix multiplication.

def multiply_matrices(matrix_a, matrix_b):
    rows_a = len(matrix_a)
    cols_a = len(matrix_a[0])
    rows_b = len(matrix_b)
    cols_b = len(matrix_b[0])

    # Check if matrices can be multiplied
    if cols_a != rows_b:
        print("Matrices cannot be multiplied: incompatible dimensions.")
        return None

    # Initialize result matrix with zeros
    result = [[0 for _ in range(cols_b)] for _ in range(rows_a)]

    # Perform multiplication
    for i in range(rows_a):
        for j in range(cols_b):
            for k in range(cols_a):
                result[i][j] += matrix_a[i][k] * matrix_b[k][j]

    return result


matrix_a = [
    [1, 2, 3],
    [4, 5, 6]
]
matrix_b = [
    [7, 8],
    [9, 10],
    [11, 12]
]

result = multiply_matrices(matrix_a, matrix_b)
print(result)
